Pilatus new features

jordiandreu opened the pull request:
Add new features for Pilatus via tango attribute/command:

1. Ability to set/get readout geometries for Pilatus 6M model via tango command.
2. Export the camera status as tango attribute (cam_state).
3. Fix bug in trigger delay method.
4. Fix minor typos.

jordiandreu committed:
Fix typo in PilatusInterface name

Add missed character in variable name.

jordiandreu committed:
Fix bug in attribute method

Read_trigger_delay method was still implemented via an old API
(communication object).

Adapt the method to the new API.

jordiandreu committed:
Add camera status attribute

Add a read-only attribute reporting the camera status.

jordiandreu committed:
Add sendAnyCommand DS command

Add the sendAnyCommnad method as a Tango command.

jordiandreu committed:
Add set/get readout geometries methods

Add two commands to set/get specific readout geometries.
The defined geometries for Pilatus 6M are: C2, C18 and C60
(full frame).

sebastien-petitdemange commented on camera/Pilatus.py:
```diff
@@ -214,12 +233,50 @@ def write_fill_mode(self, attr):
+    def getReadoutGeometryCmd(self):
+        if _PilatusCamera.hasRoiCapability():
+            image = _CtControl.image()
+            roi = image.getRoi()
```
Hi Jordi,

I think in your case, you want to have the Hardware roi and not the combined one (Hardware + Software).
So I think you should get the roi like:
```python
hardware_image = image.getHard()
roi = hardware_image.getRoi()
```

jordiandreu replied:
Yes, I do. I will implement it on the next commit.

jordiandreu replied:
Hi @sebastien-petitdemange , I realize that method image.getHard().getRoi() does not exists?
What method do you suggest then to use to get the hardware Roi instead of the combination of Hard and Soft Rois?

sebastien-petitdemange replied:
Actually it's getRealRoi the method I was thinking off. This method will return the hardware roi applied for the camera or an empty roi (see Roi::isempty) if no hardware roi is set (i.e: full frame C60 in your case).

jordiandreu replied:
I see. I will change the code accordingly.

sebastien-petitdemange commented on camera/Pilatus.py:
```diff
@@ -214,12 +233,50 @@ def write_fill_mode(self, attr):
+        if _PilatusCamera.hasRoiCapability():
+            image = _CtControl.image()
+            roi = image.getRoi()
+            rmode = _getDictKey(self.__ReadoutRoi, roi)
```
Don't you think we should catch the KeyError in case the roi it's not yet managed and send either "UNKNOWN" or something else?

jordiandreu replied:
In fact, the exception is cached in the _getDictValue as a ValueError exception.
In case of _getDictKey method, a KeyError is cached.
In bot cases, a None is returned.

sebastien-petitdemange replied:
OK then. you can leave it like that.

sebastien-petitdemange commented on camera/Pilatus.py:
```diff
@@ -231,23 +288,29 @@ class PilatusClass(PyTango.DeviceClass):
+        'setReadoutGeometryCmd':
+        [[PyTango.DevString, "Set readout geometry"],
+         [PyTango.DevVoid, "None"]],
+        'getReadoutGeometryCmd':
```
If you change getReadoutGeometryCmd setReadoutGeometryCmd to an attribute, then in spec we'll have enumeration introspection with getAttrStringValueList...

jordiandreu replied:
You mean just define the ReadoutGeometry attribute implementing the same logic in the write/read methods that is implemented in the set/get commands?

sebastien-petitdemange replied:
exactly.

jordiandreu replied:
OK. I'll do it like this.

sebastien-petitdemange reviewed (changes requested):
Thanks for your work,

SEB

jordiandreu committed:
Implement readout_geometry as attribute

Implement the readout_geometry selection as R/W attribute.

The write method sets both the hard and soft ROIs and the read
method gathers the hard Roi. The Roi dictionary defines the full
frame geometry as the default ROI value.
